### PR TITLE
Port scalaz8 IO to 7.2.x in new module

### PR DESCRIPTION
**Pull request by jmcardon**

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -96,6 +96,22 @@ lazy val concurrent = Project(
...
 )
 
+lazy val ioeffect = Project(
+  id = "ioeffect",
+  base = file("ioeffect")
+).settings(
+  standardSettings,
+  name := IOEffectName,
+  osgiExport("scalaz.ioeffect"),
+  OsgiKeys.importPackage := Seq("javax.swing;resolution:=optional", "*"),
```
Comment (hrhino) [inline build.sbt]: wait what `javax.swing` what how why?

File: build.sbt [hunk 2]
```diff
@@ -96,6 +96,22 @@ lazy val concurrent = Project(
...
+  libraryDependencies ++=
+  Seq("org.specs2" %% "specs2-core"          % "4.0.0" % "test",
+    "org.specs2" %% "specs2-matcher-extra" % "4.0.0" % "test"),
```
Comment (xuwei-k) [inline build.sbt]: specs2 4.x does not support Scala 2.10 (and Java 7?). but scalaz 7.2.x support Scala 2.10 and Java 7.

File: build.sbt (hunk 1 above)
Comment (emilypi) [reply build.sbt]: @hrhino [https://imgflip.com/i/28029b](because)

**Comment by fommil** [comment]
This is awesome to see! I haven't done a review yet, I will try to find the time but please don't be blocked by me.

I was thinking perhaps it would make sense in a separate repo/artifact so that we don't have to make a scalaz 7.2 release every time it is updated, and also doesn't come with bincompat guarantees. I think if we broke bincompat at this stage of 7.2.x it would be very embarrassing for scalaz even if it's in a new module.

**Comment by xuwei-k** [comment]
> I was thinking perhaps it would make sense in a separate repo/artifact so that we don't have to make a scalaz 7.2 release every time it is updated, and also doesn't come with bincompat guarantees.

👍

**Comment by jdegoes** [comment]
This is great! But: do you want a couple bug fixes and the new bifunctor design (`IO[E, A]`)? Nearly ready to submit that PR to 8.x.

I know a lot of folks are looking forward to using this with 7.x.

**Comment by emilypi** [comment]
@jdegoes 
for sure, send 'em our way

File: ioeffect/src/main/scala/scalaz/ioeffect/IOInstances.scala [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+//    override final def ap[A, B](ma: IO[A])(mf: IO[A => B]): IO[B] =
+//      ma.flatMap(a => mf.map(f => f(a)))
+
+
+
+
+//////    override final def pure[A](a: A): IO[A] = IO.now(a)
+////
+////    override final def flatMap[A, B](ma: IO[A])(f: A => IO[B]): IO[B] =
+//      ma.flatMap(f)
```
Comment (xuwei-k) [inline ioeffect/src/main/scala/scalaz/ioeffect/IOInstances.scala]: unnecessary comments?

**Comment by fommil** [comment]
@jdegoes can you please set up a repo `scalaz/backports` or `scalaz/ioeffect` so we can continue development there and close this PR?

**Comment by Milyardo** [comment]
Should `effect` be deleted from 7.3 if `ioeffect` gets published as a separate project?

**Comment by fommil** [comment]
@milyardo, no it will be in 7.3. This is for 7.2.

**Comment by fommil** [comment]
@Milyardo I don't know if deprecation is the right way to go. Many basic uses of IO will remain the same. But if it is a dramatic change, and there is a `scalaz-ioeffect` package available to use in 7.2, I'd be happy with adding deprecation warnings.

**Comment by fommil** [comment]
@jdegoes thanks! Not to be a total PITA but do you mind changing it to `scalaz-ioeffect` to make it nicer when forked?

**Comment by jdegoes** [comment]
@fommil Your wish is my command.

**Pull request closed without merging**